drivers: st7789h2: parallel gpio bitbang driver

lochej opened the pull request:
Fixes #39289

Adds support for STM32L562E-DK driver TFT color display based on the ST7789H2 driver.
The initial implementation of the driver relies on GPIO bit-banging to drive the ST7789H2 with a 16bit parallel interface.
This implementation is platform generic.
An additional implementation will come and aim at using the FMC peripheral to make the driving of the LCD handled by the hardware, like ST intended.

lochej committed:
display: st7789h2: added gpio bitbang driver

Added st7789h2 gpio bitbanging driver.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

erwango reviewed (commented):
Mostly some comments as a first step, but looks good in general

gmarull reviewed (changes requested):
Please split SRAM support from display driver.

I'm not convinced that a display driver based on GPIO bit banging is useful when the hardware has a dedicated block to drive such displays efficiently.

erwango commented:
> But it takes quite some effort to configure correctly especially for this Secure/Non-Secure STM32L5 with variable base adresses.

I don't think this is required.
Main CM33 usage in Zephyr will be to use TFM as the Secure application. Everything that doesn't fit into TFM, such as vast majority of peripherals (such as display and related RAM) is handled on the non secure side and doesn't require specific register translation.
By the way this is usual to minimize the number of peripherals/stacks/drivers on secure side, and put only the sensitive parts on secure side. This is a way to minimize the solution complexity, surface of attack and then globally have a more secure solution.

mbolivar-nordic reviewed (changes requested):
Some notes and questions on the DT / driver model bits.

tejlmand commented on drivers/display/Kconfig.st7789h2:
```diff
@@ -0,0 +1,44 @@
+	  Specify the color pixel format for the ST7789H2 display controller.
+
+config ST7789H2_RGB565
+	bool "RGB565"
```
Could you please add a `help` and describe there what RGB565 means / refers to.

I see that for now there is only one possible choice value, but as you have already made this a choice, then I assume more options will follow in the future, in which case describing each option carefully is important for other users to understand.

lochej replied:
Thanks for that comment, yes in fact my implementation support only RGB565 16 bits since I had just the stm32l562e-dk and that's a 16bits bus that's routed on the board. 
Either I do it simpler and remove the "choice" or I could support 12 bits RGB444 fairly easily.

lochej replied:
Fixed this by removing this config prompt and providing support to RG565 only. RGB444 is not available through the display device API so the only meaningful ones would have been RGB888 24 bits instead but the ST7789 only supports up to 18bits pixel data so I didn't feel like spending time on this 🙃

tejlmand commented on boards/arm/stm32l562e_dk/Kconfig.defconfig:
```diff
@@ -45,4 +46,74 @@ config USE_DT_CODE_PARTITION
+
+# Original orientation of the display needs to invert Y only
+
+config LVGL_POINTER_KSCAN_SWAP_XY
+	default n
```
Why this ?

This is already a `default n` here:
https://github.com/zephyrproject-rtos/zephyr/blob/9c74027a7bbdc7348441ef2144a8c198c4e13d45/lib/gui/lvgl/Kconfig.input#L27-L29

lochej replied:
I must admit that this is probably some remains of when I ported the tft shield to the ST board, I had to disable this flag to properly align the axis of the touch controller. 

Is that a bad idea to leave it in the board def to show the user/board maintainers that swapping should be disabled ?

lochej replied:
I removed these config values since it is already default in LVGL

tejlmand commented on boards/arm/stm32l562e_dk/Kconfig.defconfig:
```diff
@@ -45,4 +46,74 @@ config USE_DT_CODE_PARTITION
+config LVGL_POINTER_KSCAN_SWAP_XY
+	default n
+
+config LVGL_POINTER_KSCAN_INVERT_X
+	default n
```
Why this ?

This is already a `default n` here:
https://github.com/zephyrproject-rtos/zephyr/blob/9c74027a7bbdc7348441ef2144a8c198c4e13d45/lib/gui/lvgl/Kconfig.input#L33-L35

lochej replied:
That's the same reason as your previous comment :)

lochej replied:
Based on your suggestions, I have removed these config lines and rely on the default values of the LVGL library.

tejlmand commented on boards/arm/stm32l562e_dk/Kconfig.defconfig:
```diff
@@ -45,4 +46,68 @@ config USE_DT_CODE_PARTITION
+	default LVGL_COLOR_DEPTH_16
+endchoice
+
+config KSCAN
+	default y
```
@lochej did you take a look at my previous comment to see if the use-case I described is working correctly with this PR ?

lochej replied:
I'm not sure I get what you mean, in the latest update, I separated the prompts for LVGL and KSCAN in the defconfig file so they could be enables separately. But if one wants to use LVGL without kscan, that should be possible. 

If you have any clues on how to improve the definition to make sure menuconfig works correctly please tell me :)

tejlmand replied:
> But if one wants to use LVGL without kscan, that should be possible.

Ok, if that's possible, then everything seems fine.

tejlmand reviewed (changes requested):
some minor observations.

mbolivar-nordic commented on boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi:
```diff
@@ -101,6 +101,14 @@
 	};
+
+	/* Display touch panel controller FT6236 is compatible with FT5336*/
+	touch_controller: ft5336@38 {
```
> I'm not sure I get what you mean, the touch_controller/FT5336 peripheral is placed under &i2c1 bus node and the reg value is its I2C address like the other peripheral on the bus LSM6DSO.

Sorry, I got thrown off by GitHub's indentation here. You are right of course.

mbolivar-nordic commented on dts/bindings/display/sitronix,st7789h2-gpio.yaml:
```diff
@@ -0,0 +1,172 @@
+    cmd-data-gpios:
+      type: phandle-array
+      required: true
+      description: D/CX pin.
```
>  The regular st7789v driver from which I copied from is not compliant though.

Thanks for reporting. It's a fun game of whack-a-mole to keep things clean :). https://github.com/zephyrproject-rtos/zephyr/pull/40234

mbolivar-nordic commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,608 @@
+	(void)data;
+	(void)config;
+
+	if (config->reset_gpio.port) {
```
One thing is missing in this function: each GPIO port should be checked with `device_is_ready` before passing the DT spec to gpio_pin_configure_dt. If device_is_ready returns false for any required devices, you should fail with -ENODEV.

Otherwise my comments are all addressed, thank you for your patience.

lochej replied:
That has been fixed, now I check for device_is_ready before configuring :-)

mbolivar-nordic commented on samples/drivers/kscan_touch/src/main.c:
```diff
@@ -19,6 +19,12 @@ LOG_MODULE_REGISTER(main);
 
 const struct device *kscan_dev = DEVICE_DT_GET(TOUCH_CONTROLLER_NODE);
 
+#ifdef CONFIG_BOARD_STM32L562E_DK
```
Introducing board-specific hacks into generic samples is a no-no.

I suggest you do something like add some optional /zephyr,user properties that configure whatever quirks you need to manage here instead.

lochej replied:
I agree with that, I don't like to have enable the LCD/touchscreen power in the kscan touch sample.
Normally, it is enabled by the display driver, but since the power lines are shared between the touch and the lcd controller, I'm a bit unsure on how to handle this.

I put the priority on the display driver since its configuration is relying on powering up the display and waiting enough time. 
Overwise, the only thing I see is that touch kscan driver needs a new device-tree property to also add an optionnal power-gpio to it and enable it in the init routine of the kscan device. Then we'll have to handle the fact that the power-gpio will be manipulated by both the display and touch driver at runtime.

lochej replied:
Fixed in latest commit. I also updated the FT5336 driver to the latest GPIO_DT api. There are no more traces of the board specific changes in the kscan driver.

mbolivar-nordic reviewed (changes requested):
DT/driver model bits look good in the driver itself, thank you. One problem with the sample remains.

I'm not sure from a testing perspective if having the sample is enough to say we are build testing the driver -- does anyone know? @galak -- do we have a policy?

mbolivar-nordic commented on dts/bindings/kscan/focaltech,ft5336.yaml:
```diff
@@ -11,3 +11,15 @@ properties:
+        POWER pin.
+
+        Enable power for the circuitry of the touch-controller.
+        Boards like stm32l562e_dk needs it.
```
Board-specific notes in generic bindings are a no-no, too :). Can you please just remove this line?

Also, heads-up that @jfischer-no has a PR going in parallel with yours which will cause conflicts for whoever gets merged second: https://github.com/zephyrproject-rtos/zephyr/pull/40765/

lochej replied:
Alright I'll remove that line. @jfischer-no 's PR looks more advanced and improves abstraction, it's ok to being merged after his PR, I'll adapt this one :)

mbolivar-nordic replied:
Thanks for your continued patience!

mbolivar-nordic reviewed (dismissed):
I gave this a quick skim and my issues seem to be addressed. Thanks for the rebase. The DT bits LGTM.

lochej commented:
> I gave this a quick skim and my issues seem to be addressed. Thanks for the rebase. The DT bits LGTM.

Thank you for this latest review 👍 It is all good now :)

jfischer-no commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,645 @@
+
+static void st7789h2_exit_sleep(const struct device *dev)
+{
+	st7789h2_transmit(dev, ST7789H2_CMD_SLEEP_OUT, NULL, 0);
```
Cannot `st7789h2_transmit()` subsequent calling `gpio_pin_set_dt()` fail?

lochej replied:
Not sure I got what you mean, can you be more precise here ?
To enter or exit sleep on the display, only a command through the bus is necessary.

jfischer-no reviewed (changes requested):
drivers/display/display_st7789h2.h header is redundant

tejlmand reviewed (dismissed):
lgtm

lochej commented:
Dear all,

I've been quite busy and I left this PR on the side, but I've rebased to support the latest LVGL v8 and FT5336 updates.
I hope you can give it another look 👍

lochej commented on drivers/display/display_st7789h2.h:
```diff
@@ -0,0 +1,75 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ST7789H2_DISPLAY_DRIVER_H__
```
It's pretty much possible yes, though I wanted to decouple the 2 drivers, so if it is mandatory to merge the 2 st7789h2 and st7789v, then I'll do it.

jfischer-no replied:
It not it almost the same controller with different interfaces? Why should we have two headers describing the same register set?

tejlmand commented on drivers/display/Kconfig.st7789h2:
```diff
@@ -0,0 +1,32 @@
+	  Enable driver for ST7789H2 display driver.
+
+config ST7789H2_GPIO_PORT_WRITING
+	bool "GPIO_PORT_WRITING"
```
Please give this a proper title and not just part of the settings name.

lochej replied:
Sounds OK to me, there was the help section describing what it does but I'll provide a better title for this 👍

tejlmand commented on drivers/kscan/Kconfig.ft5336:
```diff
@@ -24,4 +24,10 @@ config KSCAN_FT5336_INTERRUPT
 	  Enable interrupt support (requires GPIO).
 
+config KSCAN_FT5336_POWER_GPIO
+	bool "Enable Power pin"
```
As per the guidelines, please remove <word>:
https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#prompt-strings

> For a Kconfig symbol that enables a driver/subsystem FOO, consider having just “Foo” as the prompt, instead of “Enable Foo support” or the like.

Maybe just write:
```suggestion
	bool "Power pin wake-up"
```

tejlmand reviewed (commented):
couple of minor nits.

lochej commented on drivers/kscan/Kconfig.ft5336:
```diff
@@ -24,4 +24,10 @@ config KSCAN_FT5336_INTERRUPT
+	bool "Enable Power pin"
+	help
+	  Enable the usage of power-gpio to power up the touch controller
+	  when enabled power-gpio required in the device tree.
```
```suggestion
	  when enabled, power-gpio is required in the device tree.
```

jfischer-no commented on boards/arm/stm32l562e_dk/Kconfig.defconfig:
```diff
@@ -45,4 +46,59 @@ config USE_DT_CODE_PARTITION
 
+
+if LVGL
+
+config DISPLAY
+	default y
```
That should not be necessary.

lochej replied:
Correct I'm removing this

lochej commented:
@jfischer-no @tejlmand Your comments are addressed 👍 
@erwango @mbolivar-nordic Can you guys give a quick shot at this PR ? 😄

tejlmand reviewed (dismissed):
lgtm

jfischer-no commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,645 @@
+	(void)config;
+
+	/* Optional GPIOs are set to NULL when not used */
+	if (device_is_ready(config->reset_gpio.port)) {
+		if (gpio_pin_configure_dt(&config->reset_gpio, GPIO_OUTPUT_ACTIVE)) {
+			LOG_ERR("Couldn't configure reset pin");
+			return -EIO;
+		}
+	} else if (config->reset_gpio.port) {
+		LOG_ERR("reset_gpio is not ready");
+		return -ENODEV;
+	}
```
Why not just
```c
	if (config->reset_gpio.port) {
		if (!device_is_ready(config->reset_gpio.port)) {
			return -ENODEV;
		} 

		if (gpio_pin_configure_dt(&config->reset_gpio, GPIO_OUTPUT_ACTIVE)) {
			LOG_ERR("Couldn't configure reset pin");
			return -EIO;
		}
	}
```
?

lochej replied:
Because "reset_gpio" is an optionnal dt-node for this device. With this code if device_is_ready fails, it's either a **NULL pointer on the port** or a **port not ready**.
The else if is there to catch the **port not ready** error. But the code should pass in the case of a **NULL pointer on the port** since this node is not required in the DT.

If you have a better syntax for this use case, please feel free to propose 👍

lochej replied:
Sorry I didn't read your proposition properly, it should actually be OK to implement your proposition.

lochej replied:
Done

jfischer-no commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,645 @@
+			     const enum display_pixel_format pixel_format)
+{
+
+	if (pixel_format == PIXEL_FORMAT_RGB_565) {
+		return 0;
+	}
+	LOG_ERR("Pixel format change not implemented");
+	return -ENOTSUP;
```
```c
	if (pixel_format == PIXEL_FORMAT_RGB_565) {
		return 0;
	}

	LOG_WRN("Pixel format not supported");

	return -ENOTSUP;
```

lochej replied:
Alright, didn't expect that to be a warn since I returned a ENOTSUP error, but if that's the recommended LOG_WRN for this purpose I'll include it.

Is ERR reserved for fatal blocking errors ?

jfischer-no replied:
LOG_WRN here is probably arguable. It is legitimate for a subsystem to try out what is supported, error in this case can be confusing for the user. I do not insist on LOG_WRN.

lochej replied:
Done

lochej commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,645 @@
+
+static void st7789h2_lcd_init(const struct device *dev)
+{
+	uint8_t tmp;
+
+	struct st7789h2_data *data = (struct st7789h2_data *)dev->data;
+	const struct st7789h2_config *config = (const struct st7789h2_config *)dev->config;
```
No problem with this point, I'll include this fix 👍

lochej commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,645 @@
+{
+	struct st7789h2_data *data = (struct st7789h2_data *)dev->data;
+	const struct st7789h2_config *config = (const struct st7789h2_config *)dev->config;
+	(void)data;
+	(void)config;
```
It's actually a fix for the C "Unused Variable" warning that breaks CI/UnitTests. Because "data" is used only if **CONFIG_ST7789H2_GPIO_PORT_WRITING** is enabled, then data is used, if not, Unused Variable warning will fail CI. 
Agreed that **(void) config** should be removed though because it's always used.

jfischer-no replied:
The please move data definition to `CONFIG_ST7789H2_GPIO_PORT_WRITING` block to the line above `int16_t lastpin = 0;`

lochej commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,645 @@
+
+#ifdef CONFIG_ST7789H2_GPIO_PORT_WRITING
+
+	data->port_groups_count = -1;
+	int16_t lastpin = 0;
+
+	const struct device *dev_cur = NULL;
+	const struct device *dev_pin = NULL;
```
Fair enough, I'll reorder these

lochej replied:
Done

lochej commented on drivers/kscan/kscan_ft5336.c:
```diff
@@ -179,18 +184,36 @@ static int ft5336_init(const struct device *dev)
 	struct ft5336_data *data = dev->data;
+	int r;
+
+	(void) r;
```
This kind of "unused variable" silencing is part of the Zephyr codebase here and there though.

lochej commented:
Rebased with latest LVGL changes 👍

tejlmand reviewed (dismissed):
lgtm - re-approved

lochej commented on boards/arm/stm32l562e_dk/Kconfig.defconfig:
```diff
@@ -1,6 +1,7 @@
 # STM32L562E-DK Discovery board configuration
 
 # Copyright (c) 2020 Yestin Sun
+# Copyright (c) 2021 Jeremy LOCHE
```
That's ok to me to split it into it's own commit if that's necessary. This board is pretty much the reason why this driver came up 😄

erwango replied:
ok, go ahead then

lochej replied:
Done 👍

erwango commented:
@gmarull Seems you have pending comments. Would you mind having a look?

tejlmand reviewed (dismissed):
lgtm - re-approved

gmarull commented on drivers/display/Kconfig.st7789h2:
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) 2021, Jeremy LOCHE
+# SPDX-License-Identifier: Apache-2.0
+
+config ST7789H2
+	bool "ST7789H2 display driver"
+	depends on GPIO
```
please default based on `dt_compat_enabled`

lochej replied:
Could you please point me the right syntax for that ?

lochej replied:
Done

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+#include <sys/byteorder.h>
+#include <drivers/display.h>
+
+#define LOG_LEVEL CONFIG_DISPLAY_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(display_st7789h2);
```
```suggestion
#include <logging/log.h>
LOG_MODULE_REGISTER(display_st7789h2, CONFIG_DISPLAY_LOG_LEVEL);
```

lochej replied:
Done

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+	const struct gpio_dt_spec *data_gpios;
+	const uint8_t ndata_gpios;
+
+	uint8_t vcom[1];
+	uint8_t gctrl[1];
+	uint8_t vdvvrhen;
+	uint8_t vrhs[1];
+	uint8_t vdvs[1];
+	uint8_t mdac[1];
+	uint8_t lcm[1];
+	uint8_t colmod[1];
```
what is the reason for 1 item arrays?

lochej replied:
Inspired from st7789v2 driver

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+		shift = data->port_groups[i].len;
+		offset = data->port_groups[i].offset;
+
+		valmask = ((1<<shift) - 1);
```
```suggestion
		valmask = ((1 << shift) - 1);
```
checkpatch seems to be busy with other stuff lately...

lochej replied:
Alright, I'll add this

lochej replied:
Done

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+	const struct st7789h2_config *config = dev->config;
+
+	for (uint8_t i = 0; i < config->ndata_gpios; i += 4) {
+		gpio_pin_set_dt(&(config->data_gpios[i]), value&0x1);
+		gpio_pin_set_dt(&(config->data_gpios[i+1]), value&0x2);
+		gpio_pin_set_dt(&(config->data_gpios[i+2]), value&0x4);
+		gpio_pin_set_dt(&(config->data_gpios[i+3]), value&0x8);
```
```suggestion
		gpio_pin_set_dt(&(config->data_gpios[i]), value & 0x1);
		gpio_pin_set_dt(&(config->data_gpios[i+1]), value & 0x2);
		gpio_pin_set_dt(&(config->data_gpios[i+2]), value & 0x4);
		gpio_pin_set_dt(&(config->data_gpios[i+3]), value & 0x8);
```

lochej replied:
Done

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+	LOG_DBG("Resetting display");
+
+	if (config->reset_gpio.port) {
+		k_sleep(K_MSEC(1));
+		gpio_pin_set_dt(&config->reset_gpio, 1);
+		k_sleep(K_MSEC(6));
+		gpio_pin_set_dt(&config->reset_gpio, 0);
+		k_sleep(K_MSEC(20));
+	} else {
+		st7789h2_transmit(dev, ST7789V_CMD_SW_RESET, NULL, 0);
+		k_sleep(K_MSEC(5));
```
please define or document constants, why 1, 6, 20, 5 ms?

lochej replied:
Picked from st7789v2 driver where it is not documented

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+	st7789h2_set_cmd(config, 0);
+
+	uint16_t *buffer = (void *)buf;
+	size_t len = desc->buf_size/ST7789H2_PIXEL_SIZE;
```
```suggestion
	size_t len = desc->buf_size / ST7789H2_PIXEL_SIZE;
```

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+	st7789h2_set_cs(config, 1);
+	st7789h2_set_cmd(config, 0);
+
+	uint16_t *buffer = (void *)buf;
```
```suggestion
	const uint16_t *buffer = buf;
```
?

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+	st7789h2_transmit(dev, ST7789V_CMD_PORCTRL, (uint8_t *) config->porch_param,
+			 sizeof(config->porch_param));
+
+	/* Digital Gamma Enable, default disabled */
+	tmp = 0x00;
+	st7789h2_transmit(dev, ST7789V_CMD_DGMEN, &tmp, 1);
+
+	/* Frame Rate Control in Normal Mode, default value */
+	tmp = 0x1e; /* 40 Hz */
+	st7789h2_transmit(dev, ST7789V_CMD_FRCTRL2, &tmp, 1);
```
why can't these be configured via DT?

lochej replied:
Inspired from ST7789V2 which didn't make them settable

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+	}
+	data->port_groups_count++;
+
+	LOG_INF("data_gpios optimized");
```
```suggestion
	LOG_DBG("data_gpios optimized");
```

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+static void st7789h2_exit_sleep(const struct device *dev)
+{
+	st7789h2_transmit(dev, ST7789V_CMD_SLEEP_OUT, NULL, 0);
+	k_sleep(K_MSEC(120));
```
constant should be defined or documented

lochej replied:
Actually picked from the st7789v2 driver in which it is undocumented

lochej replied:
Done, now documented

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+	.set_orientation = st7789h2_set_orientation,
+};
+
+#ifdef CONFIG_ST7789H2_GPIO_PORT_WRITING
+#define ST7789H2_DATA_INIT(inst) \
+	struct st7789h2_gpio_port_group                                                            \
+		st7789h2_data_port_groups_ ## inst[ST7789H2_DATA_NGPIOS(inst)];                    \
+	static struct st7789h2_data st7789h2_data_ ## inst = {                                     \
+		.width = DT_INST_PROP(inst, width),                                                \
+		.height = DT_INST_PROP(inst, height),                                              \
+		.x_offset = DT_INST_PROP(inst, x_offset),                                          \
+		.y_offset = DT_INST_PROP(inst, y_offset),                                          \
+		.port_groups = st7789h2_data_port_groups_ ## inst,                                 \
+		.port_groups_count = ST7789H2_DATA_NGPIOS(inst),                                   \
+	};
+#else
+#define ST7789H2_DATA_INIT(inst) \
+	static struct st7789h2_data st7789h2_data_ ## inst = {                                     \
+		.width = DT_INST_PROP(inst, width),                                                \
+		.height = DT_INST_PROP(inst, height),                                              \
+		.x_offset = DT_INST_PROP(inst, x_offset),                                          \
+		.y_offset = DT_INST_PROP(inst, y_offset),                                          \
+	};
+#endif
```
```suggestion
#ifdef CONFIG_ST7789H2_GPIO_PORT_WRITING
#define PORT_GROUPS_DEFINE(inst)                                                                   \
	static struct st7789h2_gpio_port_group                                                     \
		st7789h2_data_port_groups_ ## inst[ST7789H2_DATA_NGPIOS(inst)]                     \
#define PORT_GROUPS_INIT(inst)                                                                     \
		.port_groups = st7789h2_data_port_groups_ ## inst,                                 \
		.port_groups_count = ST7789H2_DATA_NGPIOS(inst),
#else
#define PORT_GROUPS_DEFINE(inst)
#define PORT_GROUPS_INIT(inst)
#endif

#define ST7789H2_DATA_INIT(inst)                                                                   \
	PORT_GROUPS_DEFINE(inst);                                                                  \
	static struct st7789h2_data st7789h2_data_ ## inst = {                                     \
		.width = DT_INST_PROP(inst, width),                                                \
		.height = DT_INST_PROP(inst, height),                                              \
		.x_offset = DT_INST_PROP(inst, x_offset),                                          \
		.y_offset = DT_INST_PROP(inst, y_offset),                                          \
		PORT_GROUPS_INIT(inst),                                                            \
	};
```

(and possibly just drop ST7789H2_DATA_INIT/ST7789H2_CONFIG_INIT, and put its content directly in the definition macro)

lochej replied:
Nice alternative thanks :)

lochej replied:
Unfortunately, I tried that but it broke the other macros so I'll stick to my solution for now

gmarull commented on drivers/display/display_st7789h2.c:
```diff
@@ -0,0 +1,647 @@
+
+#define ST7789H2_CONFIG_INIT(inst) \
+	static const struct gpio_dt_spec st7789h2_config_data_gpios_ ## inst[] = {                 \
+		UTIL_LISTIFY(ST7789H2_DATA_NGPIOS(inst), _DATA_GPIO_ITEM_CFG_INIT, inst) };        \
```
`UTIL_LISTIFY` is deprecated, please use `LISTIFY`

lochej replied:
Done

gmarull commented on drivers/kscan/kscan_ft5336.c:
```diff
@@ -179,18 +184,36 @@ static int ft5336_init(const struct device *dev)
 		return -ENODEV;
 	}
 
+#ifdef CONFIG_KSCAN_FT5336_POWER_GPIO
+
+	if (!device_is_ready(config->power_gpio.port)) {
+		LOG_ERR("power_gpio is not ready");
+		return -ENODEV;
+	}
+
+	r = gpio_pin_configure_dt(&(config->power_gpio),
+			GPIO_OUTPUT | GPIO_ACTIVE_LOW | GPIO_OUTPUT_ACTIVE);
+	if (r < 0) {
+		LOG_ERR("Could not configure configure power GPIO pin");
+		return r;
+	}
+
+#endif
```
you can use a regulator or power domains to achieve that without the need to modify the driver itself. See e.g. `boards/arm/thingy53_nrf5340/thingy53_nrf5340_common.dts`

lochej replied:
Interesting, that might also be useful for the display driver

lochej replied:
Done

gmarull commented on dts/bindings/display/sitronix,st7789h2-gpio.yaml:
```diff
@@ -0,0 +1,180 @@
+
+compatible: "sitronix,st7789h2-gpio"
+
+include: base.yaml
```
please also inherit from `display-controller` and re-use properties

lochej replied:
Done

lochej commented:
@erwango @gmarull @tejlmand This PR is taking a while and a lot of reviews, sorry about that, but @gmarull 's comments changed the game a little and I could simplify and refactor a good chunk of the driver. I'll have to cleanup the commits to keep them minimal but if you guys have comments before I do that I'll include them 👍

lochej committed:
boards: stm32l562e_dk: added st7789h2 support

Added device tree definitions to enable support for
ST7789H2 gpio bitbang display driver and touch controller.

Signed-off-by: Jeremy LOCHE <lochejeremy@gmail.com>

erwango commented:
@lochej Would you mind rebasing (as this is anyway required, and if done after the review, we'll need to approve again)?
Don't hesitate to ping us once it's done.

lochej commented:
@erwango @gmarull Rebasing is done :) it is ready for review !

lochej commented:
@gmarull @erwango Could you guys review this PR ? It is ready unless you have some comments.

mbolivar-nordic commented:
@gmarull ping?

gmarull commented on drivers/display/display_st7789h2_gpio.c:
```diff
@@ -0,0 +1,622 @@
+
+#include "display_st7789v.h"
+
+#include <pm/device.h>
```
@lochej please update headers with `<zephyr/` prefix.

lochej replied:
done

gmarull commented:
@lochej does this driver apply now that https://github.com/zephyrproject-rtos/zephyr/pull/44448 is in place?

lochej commented:
> @lochej does this driver apply now that #44448 is in place?

This driver is platform independant so it could be useful to other socs. Though, I can replace/update support for a new PR that could work with FMC on the stm32l562e_dk.

The thing is that no other boards use this gpio-bitbang driver for now. And once it's replaced by the FMC variant. The question is does it make sense that we merge this one meanwhile the FMC driver is being developped ?
